### ISSUE TITLE
Question regarding loss computation in multi gpu training

### ISSUE THREAD

**Issue by ratishsp** [OpenNMT/OpenNMT-py#2022]
Hi,
I am going through the code when executed in a multi-gpu setting. And I believe that the current loss computation is not correct.
Let me explain why.
As I understand, in multi-gpu training, the normalization is obtained by summing them across the gpus in world size. 
https://github.com/OpenNMT/OpenNMT-py/blob/2212db9ac2b913826764950cfdacdf8f7bee29a0/onmt/trainer.py#L237-L240
And, the loss is computed using the following lines:
https://github.com/OpenNMT/OpenNMT-py/blob/2212db9ac2b913826764950cfdacdf8f7bee29a0/onmt/trainer.py#L372-L379
And the gradients are updated in each gpu using the following lines of code:
https://github.com/OpenNMT/OpenNMT-py/blob/2212db9ac2b913826764950cfdacdf8f7bee29a0/onmt/trainer.py#L394-L402

My question is: Shouldn't the loss be added up considering all the gpus?
Something like:
`                loss = sum(onmt.utils.distributed
                                .all_gather_list
                                (loss))`
And then the loss be normalized and backpropagated through each gpu?

Thanks.

**Comment by ratishsp** [comment]
Hi @vince62s @francoishernandez
Any tips on this question are appreciated.

**Comment by francoishernandez** [comment]
Multi-GPU is kind of similar to gradient accumulation here, we accumulate the gradients over N GPUs (hence the normalization over the 'full'/'true' batches) and then perform the update.
There are probably better/cleaner/more readable way of doing this, maybe with pytorch `DataParallel` mechanisms for instance. This would require quite a bit of work and refactoring.

**Comment by ratishsp** [comment]
Thanks @francoishernandez 
One further question regarding the loss computation in https://github.com/OpenNMT/OpenNMT-py/blob/2212db9ac2b913826764950cfdacdf8f7bee29a0/onmt/utils/loss.py#L195-L198 

As per my understanding, this code is executed on each gpu. Thus the loss on each gpu is normalized using the (incorrect) value of normalization of the 'full'/'true' batches.
A concrete example: If the batch size = 10 and there are 4 gpus, then the `normalization=40`.
Then the loss is computed on each gpu by dividing by 40. It should have been divided by 10.

Let me know if I am missing something.

**Comment by guillaumekln** [comment]
The normalization by the full batch size is correct because the gradients coming from each GPU are added, not averaged.

**Comment by ratishsp** [comment]
Thanks @guillaumekln 
I think I get your point.
I believe the code works fine when each gpu handles batches of equal sizes.
My use case is bit unique, as in my scenario each gpu may process a batch of unequal size.
For eg: in a four gpu setup, the batch sizes in each gpu may be 2, 3, 3, 2. 
The full batch size (as per below code) is 2 + 3 + 3+ 2 = 10. 
https://github.com/OpenNMT/OpenNMT-py/blob/2212db9ac2b913826764950cfdacdf8f7bee29a0/onmt/trainer.py#L237-L240
Thus, the normalization as computed using the code will be inaccurate for loss computation on each gpu. This is because I need a different value of normalization for each gpu.

What I will need is a simpler code similar to:
`normalization = batch_size * world_size`
The normalization for each gpu (considering a 4 gpu setup) will be 8, 12, 12, 8 
This will result in correct loss computation on each gpu.

Let me know if it makes sense.
Thanks for your response.

**Comment by guillaumekln** [comment]
In fact the current implementation is accurate especially when the batch sizes are unequal, since the loss is always divided by the true global batch size. One the other hand, the normalization you are proposing is an approximation and assumes that each GPU gets the same batch size.

In other words, you want the normalization to be the same whether the full batch is run on a single GPU, or split into N parts distributed on N GPUs.

**Comment by ratishsp** [comment]
Thanks @guillaumekln 
I guess I am a bit confused now.
When you say 
> since the loss is always divided by the true global batch size

do you mean the total loss of the all the gpus combined?

I don't find any code which combines the loss of all the gpus, something like this:
`loss = sum(onmt.utils.distributed.all_gather_list(loss))`

Going through the code in [loss.py](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/utils/loss.py), I assumed that the loss in each gpu is divided by the full batch size, which seems incorrect. As ideally, the loss in each gpu should be divided by the batch size it has processed (and the number of gpus).

Let me know if it makes sense.
Thanks.

**Comment by guillaumekln** [comment]
> do you mean the total loss of the all the gpus combined?

No, I meant divided by the total batch size as computed by the code snippet you posted above:

https://github.com/OpenNMT/OpenNMT-py/blob/2212db9ac2b913826764950cfdacdf8f7bee29a0/onmt/trainer.py#L237-L240

> I assumed that the loss in each gpu is divided by the full batch size, which seems incorrect.

Why do you think this is incorrect?

You should think about what would be the normalization if you were to concatenate the batches processed on each GPU to build a single batch. In the example above the normalization on this combined batch would be 2 + 3 + 3 + 2 = 10.

**Comment by ratishsp** [comment]
Hi @guillaumekln 
I understand your point regarding normalization as being the full batch size.
What I am now unsure about is its impact on the loss computation.

Consider an example. 
Lets say there are 2 gpus. The batch sizes in the gpus are `b1` and `b2`. And the respective unnormalized losses are `loss1` and `loss2`.
Going by the code it seems we are computing `loss1/(b1 + b2)` and `loss2/(b1 + b2)`, then we invoke the method `onmt.utils.distributed.all_reduce_and_rescale_tensors`

I think the correct code for the loss computation should be: `((loss1/b1) + (loss2/b2)) * 1/2`.

Let me know if it makes sense.
Thanks.

**Comment by guillaumekln** [comment]
> I think the correct code for the loss computation should be: ((loss1/b1) + (loss2/b2)) * 1/2.

No, this is not correct. I mean in practice it will work fine because in most trainings `b1` and `b2` are very similar, but it is less accurate than the other normalization.

**Comment by ratishsp** [comment]
I get the point. Thanks for your answers!
